File: integration-test/java/src/test/java/de/tobiasblaschke/integration/midi/adapter/grammar/LexerTestUtils.java
package de.tobiasblaschke.integration.midi.adapter.grammar;

import de.tobiasblaschke.lib.midi.adapter.grammar.Midi1FileLexer;
import de.tobiasblaschke.lib.midi.adapter.grammar.Midi1FileParser;
import org.antlr.v4.runtime.*;
import org.jetbrains.annotations.Nullable;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.AbstractMap;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.IntStream;

import static de.tobiasblaschke.lib.midi.adapter.grammar.ParserUtils.charHexCodesAsString;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.junit.jupiter.api.Assertions.assertEquals;

public final class LexerTestUtils {
    private LexerTestUtils() { } // Disable instance creation

    public static CommonTokenStream lexToStream(final byte[] block) {
        try (InputStream byteStream = new ByteArrayInputStream(block)) {
            return lexToStream(byteStream);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    public static CommonTokenStream lexToStream(InputStream byteStream) {
        try {
            CharStream bytesAsChar = CharStreams.fromStream(byteStream, StandardCharsets.ISO_8859_1);
            Midi1FileLexer lexer = new Midi1FileLexer(bytesAsChar);
            return new CommonTokenStream(lexer);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

    }

    public static List<Token> lexToList(final byte[] block) {
        try (InputStream byteStream = new ByteArrayInputStream(block)) {
            return lexToList(byteStream);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    public static List<Token> lexToList(InputStream byteStream) {
        return tokenStreamAsList(lexToStream(byteStream));
    }

    public static List<Token> tokenStreamAsList(CommonTokenStream tokenStream) {
        var tokenSource = tokenStream.getTokenSource();

        var ret = new ArrayList<Token>();
        var nextToken = tokenSource.nextToken();
        while (nextToken != null && nextToken.getType() != Token.EOF) {
            ret.add(nextToken);
            nextToken = tokenSource.nextToken();
        }

        return ret;
    }

    public static void assertTokensMatch(List<Token> actual, TokenMatcher... matchers) {
        IntStream.range(0, matchers.length)
                        .mapToObj(index -> new AbstractMap.SimpleEntry<Integer, TokenMatcher>(index, matchers[index]))
                        .forEach(im -> {
                            var token = actual.get(im.getKey());
                            assertThat("Mismatch on token " + im.getKey(), token, im.getValue());
                        });

        assertEquals(matchers.length, actual.size(), "Expected number of tokens does not match number of matchers");
    }

    public static Midi1FileParser makeParser(final CommonTokenStream tokens) {
        var ret = new Midi1FileParser(tokens);
        ret.setErrorHandler(new DefaultErrorStrategy() {
            @Override
            protected String getTokenErrorDisplay(final @Nullable Token t) {
                if (t == null) return "<no token>";
                var text = t.getText();
                var hex = (text == null) ? "- no text -" : charHexCodesAsString(text, " ", 16, "    ");
                var tokenName = Midi1FileLexer.VOCABULARY.getDisplayName(t.getType());
                return String.format(
                        "ERROR: on %s = '%s' at %d (0x%02X) length %d",
                        tokenName, hex,
                        t.getStartIndex(),
                        t.getStartIndex(),
                        t.getStopIndex() - t.getStartIndex() + 1);
            }
        });

        return ret;
    }

}

File: midi-file-grammar/src/test/kotlin/de/tobiasblaschke/lib/midi/adapter/grammar/LexerTestUtils.kt
package de.tobiasblaschke.lib.midi.adapter.grammar

import de.tobiasblaschke.lib.midi.adapter.grammar.ParserUtils.charHexCodesAsString
import org.antlr.v4.runtime.*
import org.hamcrest.Description
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.TypeSafeMatcher
import java.io.InputStream
import java.nio.charset.StandardCharsets
import kotlin.test.assertEquals


object LexerTestUtils {

    fun lexToStream(block: UByteArray): CommonTokenStream =
        lexToStream(block.joinToString(separator = "") { Char(it.toInt()).toString() })

    fun lexToStream(block: String): CommonTokenStream =
        block.byteInputStream(charset = Charsets.ISO_8859_1)
            .use { lexToStream(it) }

    fun lexToStream(byteStream: InputStream): CommonTokenStream {
        val bytesAsChar = CharStreams.fromStream(byteStream, StandardCharsets.ISO_8859_1)
        //val lexer = Midi1FileLexer(bytesAsChar)
        val lexer = Midi1FileLexer(bytesAsChar)
        return CommonTokenStream(lexer)
    }

    fun lexToList(block: UByteArray): List<Token> =
        lexToList(block.joinToString(separator = "") { Char(it.toInt()).toString() })


    fun lexToList(block: String): List<Token> =
        tokenStreamAsList(lexToStream(block))

    fun lexToList(inputStream: InputStream) =
        tokenStreamAsList(lexToStream(inputStream))

    fun tokenStreamAsList(tokenStream: CommonTokenStream): List<Token> {
        val tokenSource = tokenStream.tokenSource

        val ret = mutableListOf<Token>()
        var nextToken = tokenSource.nextToken()
        while (nextToken != null && nextToken.type != Token.EOF) {
            ret.add(nextToken)
            nextToken = tokenSource.nextToken()
        }
        return ret
    }

    fun assertTokensMatch(actual: List<Token>, vararg matchers: TokenMatcher) {
        matchers
            .forEachIndexed { index, matcher ->
                val token = actual.getOrNull(index)
                assertThat("Mismatch on token $index", token, matcher)
            }

        assertEquals(matchers.size, actual.size, "Expected ${matchers.size} Tokens, got ${actual.size}")
    }

    fun Token.dump(): String {
        val stringType = Midi1FileLexer.VOCABULARY.getDisplayName(type)
        return "$stringType(${text.charHexCodesAsString()} = $text) at position $startIndex to $stopIndex"
    }

    fun makeDummyToken(type: Int, text: String, startIndex: Int = 0, stopIndex: Int = startIndex + text.length): Token =
        CommonToken(type, text)
            .also {
                it.startIndex = startIndex
                it.stopIndex = stopIndex }

    fun makeParser(tokens: CommonTokenStream) =
        Midi1FileParser(tokens)
            .also {
                it.errorHandler = object : DefaultErrorStrategy() {
                    override fun getTokenErrorDisplay(t: Token?): String {
                        if (t == null) return "<no token>"
                        val hex = t.text?.charHexCodesAsString() ?: " - no text -"
                        val tokenName = Midi1FileLexer.VOCABULARY.getDisplayName(t.type)
                        return String.format(
                            "ERROR: on $tokenName = '$hex' at %d (0x%02X) length %d",
                            t.startIndex,
                            t.startIndex,
                            t.stopIndex - t.startIndex + 1
                        )
                    }
                }
            }
}

class TokenMatcher(
    private val type: Int,
    private val text: String,
    private val channel: Int? = null) : TypeSafeMatcher<Token>() {
    override fun matchesSafely(actual: Token): Boolean {
        return try {
            (actual.type == type) &&
                    (actual.text == text) &&
                    (channel == null || actual.channel == channel)
        } catch (nfe: NumberFormatException) {
            false
        }
    }

    override fun describeTo(desc: Description) {
        val expectedType = Midi1FileLexer.VOCABULARY.getDisplayName(type)

        if (channel != null) {
            val channelText = Midi1FileLexer.channelNames[channel]
            desc.appendText("Expected $expectedType(${text.charHexCodesAsString()}) on $channelText")
        } else {
            desc.appendText("Expected $expectedType(${text.charHexCodesAsString()})")
        }
    }

    override fun describeMismatchSafely(item: Token, desc: Description) {
        val actualType = Midi1FileLexer.VOCABULARY.getDisplayName(item.type)

        if (channel != null) {
            val channelText = Midi1FileLexer.channelNames[item.channel]
            desc.appendText("Got $actualType(${item.text.charHexCodesAsString()}) on $channelText at position ${item.startIndex}")
        } else {
            desc.appendText("Got $actualType(${item.text.charHexCodesAsString()}) at position ${item.startIndex}")
        }
    }


}